module/zfs/zio.c

```diff
@@ -3245,16 +3245,36 @@ zio_ddt_write(zio_t *zio)
 	ddt_t *ddt = ddt_select(spa, bp);
 	ddt_entry_t *dde;
 	ddt_phys_t *ddp;
+	boolean_t added = B_FALSE;
 
 	ASSERT(BP_GET_DEDUP(bp));
 	ASSERT(BP_GET_CHECKSUM(bp) == zp->zp_checksum);
 	ASSERT(BP_IS_HOLE(bp) || zio->io_bp_override);
 	ASSERT(!(zio->io_bp_override && (zio->io_flags & ZIO_FLAG_RAW)));
 
 	ddt_enter(ddt);
-	dde = ddt_lookup(ddt, bp, B_TRUE);
+	if (spa->spa_dedup_max_entries &&
+	    spa->spa_dedup_entries >= spa->spa_dedup_max_entries) {
+		/* Over the desired DDT size, set 'nogrow' flag */
+		dde = ddt_lookup(ddt, bp, B_TRUE, NULL);
+	} else {
+		dde = ddt_lookup(ddt, bp, B_FALSE, &added);
+	}
+	if (dde == NULL) {
+		zp->zp_dedup = B_FALSE;
```

Comment:
might need to `zp->zp_dedup_verify = B_FASLE` as well

Reply:
You can look at the property dedupentries, and compare it to dedupentriesmax.  My test script for this filled up the ddt by creating random files, looping until dedupentries was greater than or equal to dedupentriesmax.

Reply:
Hm, don't know why that comment went *there*.

It didn't *look* like we needed to deal with zp_dedup_verify, since zp_dedup is set to B_FALSE, but I'll admit I don't think it *hurts* to set it.